### PR TITLE
Temporary Memory Manager

### PR DESCRIPTION
**Pull request by lnkuiper**
Currently, memory-intensive operations that can be offloaded to disk try to limit their memory usage to around 60% of the memory limit. This works well if there is exactly one of these operations happening at the same time. If multiple memory-intensive operations are happening simultaneously, e.g., when multiple hash tables are being probed within the same pipeline or multiple connections are running queries concurrently, their combined memory usage may exceed 100%, causing DuckDB to throw an `OutOfMemoryException.`

This PR implements `TemporaryMemoryManager`, which manages the temporary memory of concurrent operations. The `TemporaryMemoryManager` is a one-of class that lives in the `BufferPool`. A memory-intensive operation can register that it is currently active using the following function:
```c++
class TemporaryMemoryManager {
	// ...

	//! Register a TemporaryMemoryState
	unique_ptr<TemporaryMemoryState> Register(ClientContext &context);

	// ...
};
```

As long as the state is within scope, it is active. When it goes out of scope, it is automatically removed from the `TemporaryMemoryManager`'s internal state. Each `TemporaryMemoryState` is guaranteed some minimum amount of memory by the `TemporaryMemoryManager` depending on the number of threads and memory limit.

The memory-intensive operations communicate how much data they have through the API of the `TemporaryMemoryState`:
```c++
//! State of the temporary memory to be managed concurrently with other states
//! As long as this is within scope, it is active
class TemporaryMemoryState {
	friend class TemporaryMemoryManager;

private:
	TemporaryMemoryState(TemporaryMemoryManager &temporary_memory_manager, idx_t minimum_reservation);

public:
	~TemporaryMemoryState();

public:
	//! Set the remaining size needed for this state, and updates the reservation
	void SetRemainingSize(ClientContext &context, idx_t new_remaining_size);
	//! Get the remaining size that was set for this state
	idx_t GetRemainingSize() const;
	//! Set the minimum reservation for this state
	void SetMinimumReservation(idx_t new_minimum_reservation);
	//! Get the reservation of this state
	idx_t GetReservation() const;

private:
	//! The TemporaryMemoryManager that owns this state
	TemporaryMemoryManager &temporary_memory_manager;

	//! The remaining size needed if it could fit fully in memory
	atomic<idx_t> remaining_size;
	//! The minimum reservation for this state
	atomic<idx_t> minimum_reservation;
	//! How much memory this operator has reserved
	atomic<idx_t> reservation;
};
```

For example, the `PhysicalHashJoin` may indicate that it needs 10GB of memory by using `SetRemainingSize`. This will cause the `TemporaryMemoryManager` to check its internal state and update the `reservation` accordingly. If not enough memory is available, it could, for example, reserve only 1GB. In this case, the `PhysicalHashJoin` must adapt and perform a partitioned hash join instead of a full in-memory one.

The `PhysicalHashJoin` can update its `remaining_size` after each round of hash joining partitions, and the `TemporaryMemoryManager` may decide to increase or decrease the `reservation`, depending on other registered `TemporaryMemoryState`s. Again, `PhysicalHashJoin` must adapt accordingly. If, for example, the largest partition is 500MB, the `PhysicalHashJoin` may indicate that by using `SetMinimumReservation`. This guarantees that the `TemporaryMemoryManager` will not reduce the reservation below this.

If we have, for example, this fun query plan:
```
┌───────────────────────────┐                                                                                                                                                 
│      RESULT_COLLECTOR     │                                                                                                                                                 
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                                                 
│             0             │                                                                                                                                                 
│          (0.00s)          │                                                                                                                                                 
└─────────────┬─────────────┘                                                                                                                                                                              
┌─────────────┴─────────────┐                                                                                                                                                 
│    UNGROUPED_AGGREGATE    │                                                                                                                                                 
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                                                 
│          max(#0)          │                                                                                                                                                 
│          max(#1)          │                                                                                                                                                 
│          max(#2)          │                                                                                                                                                 
│          max(#3)          │                                                                                                                                                 
│          max(#4)          │                                                                                                                                                 
│          max(#5)          │                                                                                                                                                 
│          max(#6)          │                                                                                                                                                 
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                                                 
│             1             │                                                                                                                                                 
│          (1.59s)          │                                                                                                                                                 
└─────────────┬─────────────┘                                                                                                                                                                              
┌─────────────┴─────────────┐                                                                                                                                                 
│         PROJECTION        │                                                                                                                                                 
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                                                 
│             i             │                                                                                                                                                 
│             j             │                                                                                                                                                 
│             j             │                                                                                                                                                 
│             j             │                                                                                                                                                 
│             j             │                                                                                                                                                 
│             j             │                                                                                                                                                 
│             j             │                                                                                                                                                 
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                                                 
│         100000000         │                                                                                                                                                 
│          (0.16s)          │                                                                                                                                                 
└─────────────┬─────────────┘                                                                                                                                                                              
┌─────────────┴─────────────┐                                                                                                                                                 
│         HASH_JOIN         │                                                                                                                                                 
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                                                 
│           INNER           │                                                                                                                                                 
│           i = i           │                                                                                                                                                 
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐              
│       EC: 115423371       │                                                                                                                                  │              
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                                  │              
│         100000000         │                                                                                                                                  │              
│         (971.70s)         │                                                                                                                                  │              
└─────────────┬─────────────┘                                                                                                                                  │                                           
┌─────────────┴─────────────┐                                                                                                                    ┌─────────────┴─────────────┐
│         HASH_JOIN         │                                                                                                                    │         SEQ_SCAN          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           INNER           │                                                                                                                    │            tbl            │
│           i = i           │                                                                                                                    │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    │             i             │
│       EC: 112159226       ├─────────────────────────────────────────────────────────────────────────────────────────────────────┐              │             j             │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                     │              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         100000000         │                                                                                                     │              │       EC: 100000000       │
│         (1254.71s)        │                                                                                                     │              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│                           │                                                                                                     │              │         100000000         │
│                           │                                                                                                     │              │          (0.31s)          │
└─────────────┬─────────────┘                                                                                                     │              └───────────────────────────┘                             
┌─────────────┴─────────────┐                                                                                       ┌─────────────┴─────────────┐                             
│         HASH_JOIN         │                                                                                       │         SEQ_SCAN          │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                       │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│           INNER           │                                                                                       │            tbl            │                             
│           i = i           │                                                                                       │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                       │             i             │                             
│       EC: 108987390       ├────────────────────────────────────────────────────────────────────────┐              │             j             │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                        │              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│         100000000         │                                                                        │              │       EC: 100000000       │                             
│         (692.03s)         │                                                                        │              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│                           │                                                                        │              │         100000000         │                             
│                           │                                                                        │              │          (0.28s)          │                             
└─────────────┬─────────────┘                                                                        │              └───────────────────────────┘                                                          
┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐                                                          
│         HASH_JOIN         │                                                          │         SEQ_SCAN          │                                                          
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          
│           INNER           │                                                          │            tbl            │                                                          
│           i = i           │                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          │             i             │                                                          
│       EC: 105905253       ├───────────────────────────────────────────┐              │             j             │                                                          
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                           │              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          
│         100000000         │                                           │              │       EC: 100000000       │                                                          
│         (445.80s)         │                                           │              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          
│                           │                                           │              │         100000000         │                                                          
│                           │                                           │              │          (0.27s)          │                                                          
└─────────────┬─────────────┘                                           │              └───────────────────────────┘                                                                                       
┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐                                                                                       
│         HASH_JOIN         │                             │         SEQ_SCAN          │                                                                                       
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                       
│           INNER           │                             │            tbl            │                                                                                       
│           i = i           │                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                       
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             │             i             │                                                                                       
│       EC: 102910277       ├──────────────┐              │             j             │                                                                                       
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                       
│         100000000         │              │              │       EC: 100000000       │                                                                                       
│         (233.01s)         │              │              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                       
│                           │              │              │         100000000         │                                                                                       
│                           │              │              │          (0.33s)          │                                                                                       
└─────────────┬─────────────┘              │              └───────────────────────────┘                                                                                                                    
┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                                                                                    
│         SEQ_SCAN          ││         SEQ_SCAN          │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│            tbl            ││            tbl            │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│             i             ││             i             │                                                                                                                    
│             j             ││             j             │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│       EC: 100000000       ││       EC: 100000000       │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│         100000000         ││         100000000         │                                                                                                                    
│          (0.38s)          ││          (0.26s)          │                                                                                                                    
└───────────────────────────┘└───────────────────────────┘
```
We're probing 5 hash tables in the same pipeline. I've changed the order in which pipeline dependencies are added so that the deepest pipelines get priority; therefore, building the deepest hash tables happens first. The `TemporaryMemoryManager` assigns up to 60% of the remaining free memory, so if these hash tables are built first, they get a larger slice of the remaining memory. As we can see, this causes the deepest hash join to take the least time (233s), as it has a larger memory reservation and, therefore, needs to spill less data to disk.

This is desirable because the sooner the deepest hash join finishes its partitioned probing, the sooner it can fully free up memory, leaving more available for the other hash joins.

The example query plan that I show here is created by a query that looks something like this:
```sql
CREATE TABLE tbl AS
SELECT range i,
       range j
FROM range(100000000);

SELECT max(i),
       max(t1.j),
       max(t2.j),
       max(t3.j),
FROM tbl AS t1
JOIN tbl AS t2 USING (i)
JOIN tbl AS t3 USING (i);
```
This joins 3 tables, but in the example query plan, we join 6 tables.

Each hash table should take somewhere ~4GB of memory. However, in the example query plan, I have given DuckDB a memory limit of 5GB. Currently, DuckDB will throw an `OutOfMemoryException` for this query as the combined memory usage exceeds the limit. With this PR, DuckDB is now able to complete it without problems.

I've also integrated this into `RadixPartitionedHashTable` (for hash aggregation), but no other operators (yet) as of now. This is left for a future PR.

### CODE REVIEW

File: src/execution/operator/join/physical_hash_join.cpp [hunk 1]
```diff
@@ -346,7 +349,24 @@ class HashJoinPartitionEvent : public BasePipelineEvent {
...
+
+		// Assume 8 blocks per partition per thread (4 input, 4 output)
+		auto partition_multiplier =
+		    RadixPartitioning::NumberOfPartitions(sink.hash_table->GetRadixBits() - JoinHashTable::INITIAL_RADIX_BITS);
+		auto thread_memory = 8 * partition_multiplier * Storage::BLOCK_ALLOC_SIZE;
+		auto repartition_threads = MaxValue<idx_t>(sink.temporary_memory_state->GetReservation() / thread_memory, 1);
+
+		if (repartition_threads < local_hts.size()) {
+			// Limit the number of threads working on repartitioning based on our memory reservation
+			for (idx_t thread_idx = repartition_threads; thread_idx < local_hts.size(); thread_idx++) {
```
Comment (Mytherin) [inline src/execution/operator/join/physical_hash_join.cpp:362]: Is this triggered in the tests anywhere?

File: src/execution/operator/join/physical_hash_join.cpp [hunk 2]
```diff
@@ -685,8 +729,11 @@ void HashJoinGlobalSourceState::PrepareBuild(HashJoinGlobalSinkState &sink) {
 	D_ASSERT(global_stage != HashJoinSourceStage::BUILD);
 	auto &ht = *sink.hash_table;
 
+	// Update remaining size
+	sink.temporary_memory_state->SetRemainingSize(sink.context, ht.GetRemainingSize());
+
 	// Try to put the next partitions in the block collection of the HT
-	if (!sink.external || !ht.PrepareExternalFinalize()) {
+	if (!sink.external || !ht.PrepareExternalFinalize(sink.temporary_memory_state->GetReservation())) {
 		global_stage = HashJoinSourceStage::DONE;
```
Comment (Mytherin) [inline src/execution/operator/join/physical_hash_join.cpp:737]: Can you check if the temporary memory state is correctly cleaned up here?

File: src/parallel/executor.cpp [hunk 3]
```diff
@@ -243,7 +243,7 @@ void Executor::VerifyScheduledEventsInternal(const idx_t vertex, const vector<Ev
 	for (auto parent : parents) {
 		idx_t i;
 		for (i = 0; i < count; i++) {
-			if (vertices[i] == parent) {
+			if (&vertices[i].get() == &parent.get()) {
```
Comment (Mytherin) [inline src/parallel/executor.cpp:246]: nit: `RefersToSameObject`

File: src/parallel/meta_pipeline.cpp [hunk 4]
```diff
@@ -92,93 +91,93 @@ MetaPipeline &MetaPipeline::CreateChildMetaPipeline(Pipeline &current, PhysicalO
...
-	return pipelines.back().get();
+	return *pipelines.back();
 }
 
-void MetaPipeline::AddDependenciesFrom(Pipeline *dependant, Pipeline *start, bool including) {
+void MetaPipeline::AddDependenciesFrom(Pipeline &dependant, Pipeline &start, bool including) {
 	// find 'start'
 	auto it = pipelines.begin();
-	for (; it->get() != start; it++) {
+	for (; it->get() != &start; it++) {
```
Comment (Mytherin) [inline src/parallel/meta_pipeline.cpp:103]: `RefersToSameObject`

File: src/parallel/meta_pipeline.cpp [hunk 5]
```diff
@@ -92,93 +91,93 @@ MetaPipeline &MetaPipeline::CreateChildMetaPipeline(Pipeline &current, PhysicalO
...
 	if (!including) {
 		it++;
 	}
 
 	// collect pipelines that were created from then
-	vector<Pipeline *> created_pipelines;
+	vector<reference<Pipeline>> created_pipelines;
 	for (; it != pipelines.end(); it++) {
-		if (it->get() == dependant) {
+		if (it->get() == &dependant) {
```
Comment (Mytherin) [inline src/parallel/meta_pipeline.cpp:113]: `RefersToSameObject`

File: src/parallel/meta_pipeline.cpp [hunk 6]
```diff
@@ -92,93 +91,93 @@ MetaPipeline &MetaPipeline::CreateChildMetaPipeline(Pipeline &current, PhysicalO
...
 
-void MetaPipeline::AddFinishEvent(Pipeline *pipeline) {
+void MetaPipeline::AddFinishEvent(Pipeline &pipeline) {
 	D_ASSERT(finish_pipelines.find(pipeline) == finish_pipelines.end());
 	finish_pipelines.insert(pipeline);
 
 	// add all pipelines that were added since 'pipeline' was added (including 'pipeline') to the finish group
 	auto it = pipelines.begin();
-	for (; it->get() != pipeline; it++) {
+	for (; it->get() != &pipeline; it++) {
```
Comment (Mytherin) [inline src/parallel/meta_pipeline.cpp:131]: `RefersToSameObject`

File: src/storage/temporary_memory_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    : temporary_memory_manager(temporary_memory_manager_p), remaining_size(0),
+      minimum_reservation(minimum_reservation_p), reservation(0) {
+}
+
+TemporaryMemoryState::~TemporaryMemoryState() {
+	temporary_memory_manager.Unregister(*this);
+}
+
+void TemporaryMemoryState::SetRemainingSize(ClientContext &context, idx_t new_remaining_size) {
+	lock_guard<mutex> guard(temporary_memory_manager.lock);
```
Comment (Mytherin) [inline src/storage/temporary_memory_manager.cpp:20]: Maybe this can move into a function of the `TemporaryMemoryManager`

**Review by Mytherin — APPROVED** [review]
Thanks for the PR! Looks great - some minor comments otherwise this is good to go:

**Comment by lnkuiper** [comment]
Thanks for the feedback! I've implemented all of it, fingers crossed for CI

**Comment by lnkuiper** [comment]
I think this is ready to go!

**Comment by Mytherin** [comment]
Thanks! Good stuff :)

**Comment by szarnyasg** [comment]
@lnkuiper does this PR imply any changes to the performance guide's out-of-core section?
https://duckdb.org/docs/guides/performance/how-to-tune-workloads#limitations

**Comment by lnkuiper** [comment]
@szarnyasg Good question. It is currently only implemented in the hash join and aggregation operator, which should cover most cases where multi-operator plans can run out of memory. It is not implemented in any other operator (yet), so I think we should keep the disclaimer in the performance guide for now.

**Pull request merged**